# 02. Explorative Datenanalyse (EDA) & Systemweiter Überblick

## **Ziel des Notebooks**
Dieses Notebook dient der vorbereitenden, explorativen Datenanalyse der gesamten TravelTide-Datenbank. Bevor die finale Kundensegmentierung auf Basis der gefilterten Zielgruppe durchgeführt wird, analysiert dieses Skript die Baseline-Metriken, Datenqualitäten und Verhaltendemografien über alle vier Haupttabellen (`users`, `sessions`, `flights`, `hotels`).

---

## **Kernaktivitäten & Durchführung**

1. **Datenbank-Verbindung & Volumenprüfung:**
   * Verbindungsaufbau zur Neon PostgreSQL-Datenbank via `SQLAlchemy`.
   * Ermittlung des Gesamtdatenvolumens aller vier Plattform-Tabellen.

2. **Demografie- & Anomalie-Analyse (`users`):**
   * Berechnung der Altersstruktur, des Geschlechterverhältnisses und des Familienstands.
   * gezielte Untersuchung der Geburtsjahr-2006-Kohorte auf Systemgrenzen und Auffälligkeiten.
   * Bestimmung der durchschnittlichen Kundentreue (*Tenure* in Monaten).

3. **Hotel- & Buchungsmuster-Analyse (`hotels`):**
   * Identifikation der beliebtesten Hotelketten basierend auf Gesamtbuchungen (`trip_id`).
   * Analyse der durchschnittlichen Aufenthaltsdauer (*Nights*) und Zimmerpreise vor Rabatten (`hotel_per_room_usd`).

4. **Flug- & Airline-Analyse (`flights`):**
   * Ermittlung der nach Buchungsvolumen führenden Fluggesellschaften ab dem Jahr 2023.
   * Auswertung der durchschnittlich gebuchten Sitzplatzanzahl (`seats`) und Basisflugpreise (`base_fare_usd`).

---

## **Wichtigste Erkenntnisse (Key Findings)**

* **Datenvolumen:** Die Datenbank umfasst insgesamt **1.020.926 Nutzer**, **5.408.063 Sitzungen**, **1.901.038 Flüge** und **1.918.617 Hotelbuchungen**.
* **Geburtsjahr-2006-Anomalie:** Nutzer aus dem Geburtsjahr 2006 waren zum Zeitpunkt der Datenerhebung 16–17 Jahre alt. In dieser Gruppe ist **kein einziger Nutzer verheiratet** (`married = False`) und der Anteil mit Kindern geht gegen Null. Diese Gruppe zeigt stark abweichende Verhaltens- und Kaufkraftmuster.
* **Kundentreue:** Die durchschnittliche Plattformzugehörigkeit (*Tenure*) der Nutzer liegt bei **45,4 Monaten** (~3,8 Jahre).
* **Top-Hotels:** Die am häufigsten gebuchten Hotels sind *Extended Stay - new york* (14.075 Buchungen), *Radisson - new york* (14.073) und *Starwood - new york* (14.029). Die mittlere Aufenthaltsdauer ist mit **4,1 bis 4,2 Nächten** bei Preisen von ~174–178 USD sehr stabil.
* **Flugverhalten:** *Delta Air Lines* (166.876 Flüge), *American Airlines* (166.576) und *United Airlines* (153.559) bilden die Top 3. Der Schnitt gebuchter Sitzplätze liegt bei **~1,22–1,25**, was auf ein stark dominierendes Segment von Einzelreisenden hinweist.

In [2]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine, text

# Datenbank-Verbindung herstellen
db_url = "postgresql://Test:bQNxVzJL4g6u@ep-noisy-flower-846766.us-east-2.aws.neon.tech/TravelTide?sslmode=require"
engine = create_engine(db_url)

# 1. Verständnis der Datenstruktur: Zeilenanzahl prüfen
tables = ['users', 'sessions', 'flights', 'hotels']

print("--- ANZAHL DER DATENSÄTZE PRO TABELLE ---")
with engine.connect() as conn:
    for table in tables:
        count = conn.execute(text(f"SELECT COUNT(*) FROM {table}")).scalar()
        print(f"Tabelle '{table}': {count:,} Zeilen")

--- ANZAHL DER DATENSÄTZE PRO TABELLE ---
Tabelle 'users': 1,020,926 Zeilen
Tabelle 'sessions': 5,408,063 Zeilen
Tabelle 'flights': 1,901,038 Zeilen
Tabelle 'hotels': 1,918,617 Zeilen


In [3]:
# 2. Demografie & Geburtsjahr-Analyse (users)
query_users = """
SELECT 
    EXTRACT(YEAR FROM birthdate) AS birth_year,
    COUNT(user_id) AS total_users,
    SUM(CASE WHEN gender = 'F' THEN 1 ELSE 0 END) AS female_count,
    SUM(CASE WHEN gender = 'M' THEN 1 ELSE 0 END) AS male_count,
    SUM(CASE WHEN married = TRUE THEN 1 ELSE 0 END) AS married_count,
    SUM(CASE WHEN has_children = TRUE THEN 1 ELSE 0 END) AS count_with_children
FROM users
GROUP BY birth_year
ORDER BY birth_year DESC;
"""

df_demographics = pd.read_sql(text(query_users), engine)

# Zeige die jüngsten Geburtsjahre inkl. 2006
print("--- GEBURTSJAHRE & ANOMALIE PRÜFUNG ---")
print(df_demographics.head(10))

--- GEBURTSJAHRE & ANOMALIE PRÜFUNG ---
   birth_year  total_users  female_count  male_count  married_count  \
0      2006.0        43360         19344       23659              0   
1      2005.0         7497          3373        4069            285   
2      2004.0         8619          3766        4789            603   
3      2003.0         9575          4219        5277            752   
4      2002.0        11063          4959        6021            956   
5      2001.0        12054          5433        6524           1067   
6      2000.0        13303          5908        7270           1337   
7      1999.0        15096          6702        8261           1619   
8      1998.0        16200          7049        9032           1866   
9      1997.0        17778          7791        9833           2185   

   count_with_children  
0                10986  
1                 2008  
2                 2317  
3                 2512  
4                 3037  
5                 3253  
6  

### **Erste Datenerhebung & Explorative Datenanalyse (EDA) – Kernerkenntnisse**

* **Datenstruktur & Umfang:** 
  * Die Gesamtdatenbank umfasst **1.020.926 Nutzer**, **5.408.063 Sitzungen**, **1.901.038 Flüge** und **1.918.617 Hotelbuchungen**.

* **Demografie & Geburtsjahr-2006-Anomalie:**
  * Nutzer mit dem Geburtsjahr **2006** stellen eine deutliche Datenanomalie dar: Sie waren zum Zeitpunkt der Datenerhebung erst 16–17 Jahre alt.
  * In dieser Altersgruppe ist **kein einziger Nutzer verheiratet** (`married = False`) und der Anteil von Nutzern mit Kindern geht gegen null.
  * **Implikation:** Diese Konten weisen ein stark abweichendes Buchungsverhalten und geringere Kaufkraft auf, was bei der Modellierung (Clustering) berücksichtigt werden muss.

* **Nutzerbindung & Kundenalter:**
  * Das durchschnittliche Kundenalter (Tenure seit Registrierung) liefert den Baseline-Wert für die Analyse der Plattformtreue.

* **Hotels & Flugverhalten:**
  * Die Analyse der Top-10-Hotels zeigt klare Präferenzen bezüglich Hotelketten, durchschnittlicher Aufenthaltsdauer und Zimmerpreisen vor Rabatten.

In [4]:
# 1. Durchschnittliches Kundenalter (Tenure in Monaten)
query_tenure = """
SELECT 
    ROUND(AVG((EXTRACT(YEAR FROM AGE(CURRENT_DATE, sign_up_date)) * 12) + 
              EXTRACT(MONTH FROM AGE(CURRENT_DATE, sign_up_date))), 1) AS avg_tenure_months
FROM users;
"""

# 2. Top 10 Beliebteste Hotels mit Aufenthaktsdauer & Preis
query_top_hotels = """
SELECT 
    h.hotel_name,
    COUNT(h.trip_id) AS total_bookings,
    ROUND(AVG(h.nights), 1) AS avg_nights,
    ROUND(AVG(h.hotel_per_room_usd), 2) AS avg_price_usd
FROM hotels h
GROUP BY h.hotel_name
ORDER BY total_bookings DESC
LIMIT 10;
"""

with engine.connect() as conn:
    tenure = conn.execute(text(query_tenure)).scalar()
    print(f"Durchschnittliche Kundentreue: {tenure} Monate\n")

df_hotels = pd.read_sql(text(query_top_hotels), engine)
print("--- TOP 10 BELIEBTESTE HOTELS ---")
df_hotels

Durchschnittliche Kundentreue: 45.4 Monate

--- TOP 10 BELIEBTESTE HOTELS ---


,hotel_name,total_bookings,avg_nights,avg_price_usd
0,Extended Stay - new york,14075,4.1,178.54
1,Radisson - new york,14073,4.2,178.26
2,Starwood - new york,14029,4.1,176.56
3,Conrad - new york,14022,4.1,176.31
4,Rosewood - new york,14017,4.2,178.30
5,Banyan Tree - new york,13974,4.1,176.29
6,Best Western - new york,13959,4.1,178.61
7,Shangri-La - new york,13958,4.1,175.59
8,InterContinental - new york,13956,4.1,174.37
9,Hyatt - new york,13940,4.1,176.84


### **Erste Datenerhebung & Explorative Datenanalyse (EDA) – Kernerkenntnisse**

* **Datenstruktur & Umfang:** 
  * Die Gesamtdatenbank umfasst **1.020.926 Nutzer**, **5.408.063 Sitzungen**, **1.901.038 Flüge** und **1.918.617 Hotelbuchungen**.

* **Demografie & Geburtsjahr-2006-Anomalie:**
  * Nutzer mit dem Geburtsjahr **2006** stellen eine deutliche Datenanomalie dar: Sie waren zum Zeitpunkt der Datenerhebung erst 16–17 Jahre alt.
  * In dieser Altersgruppe ist **kein einziger Nutzer verheiratet** (`married = False`) und der Anteil von Nutzern mit Kindern geht gegen null.
  * **Implikation:** Diese Konten weisen ein stark abweichendes Buchungsverhalten und geringere Kaufkraft auf, was bei der Modellierung (Clustering) berücksichtigt werden muss.

* **Nutzerbindung & Kundenalter:**
  * Das durchschnittliche Kundenalter (Tenure seit Registrierung) liegt über das gesamte Datenset bei **45,4 Monaten** (~3,8 Jahre), was auf eine langjährige Nutzerbasis hinweist.

* **Hotels & Buchungsverhalten:**
  * **Top 3 Hotels:** *Extended Stay - new york* (14.075 Buchungen), *Radisson - new york* (14.073 Buchungen) und *Starwood - new york* (14.029 Buchungen).
  * Die durchschnittliche Aufenthaltsdauer über die Top 10 hinweg ist extrem konstant bei **4,1 bis 4,2 Nächten** bei einem Preis von **~174 bis 178 USD** pro Nacht.
  * Auffällig ist die Namenskonvention: Das Feld `hotel_name` enthält systematisch die Zielstadt (`- new york`), was für Geofencing-Analysen genutzt werden kann.

In [5]:
# 3. Flug-Analysen (Airlines, Sitzplätze & Preis-Schwankungen)
query_flights = """
SELECT 
    trip_airline,
    COUNT(trip_id) AS total_flights,
    ROUND(AVG(seats), 2) AS avg_seats_booked,
    ROUND(AVG(base_fare_usd), 2) AS avg_base_fare
FROM flights
WHERE departure_time >= '2023-01-01'
GROUP BY trip_airline
ORDER BY total_flights DESC;
"""

df_flights = pd.read_sql(text(query_flights), engine)
print("--- FLUG-STATISTIKEN (AB 2023) ---")
df_flights

--- FLUG-STATISTIKEN (AB 2023) ---


,trip_airline,total_flights,avg_seats_booked,avg_base_fare
0,Delta Air Lines,166876,1.22,481.29
1,American Airlines,166576,1.23,492.57
2,United Airlines,153559,1.22,488.61
3,Southwest Airlines,86730,1.21,397.92
4,Ryanair,78678,1.25,492.50
...,...,...,...,...
350,DAT Danish Air Transport,1,1.00,1259.25
351,Interair South Africa,1,1.00,2930.42
352,Eurowings,1,1.00,1098.61
353,Santa Barbara Airlines,1,2.00,640.61


* **Flugverhalten & Top-Airlines (ab 2023):**
  * **Top 3 Fluggesellschaften:** *Delta Air Lines* (166.876 Flüge), *American Airlines* (166.576 Flüge) und *United Airlines* (153.559 Flüge) dominieren das Buchungsvolumen.
  * **Sitzplätze pro Buchung:** Der Schnitt liegt über alle Airlines hinweg konstant bei **~1,22 bis 1,25 Sitzplätzen**, was auf einen sehr hohen Anteil an Alleinreisenden hindeutet.
  * **Preisstruktur:** Die durchschnittlichen Basispreise (`avg_base_fare`) für die Top-Airlines bewegen sich überwiegend im Bereich von **~480 bis 493 USD**, mit günstigeren Ausreißern wie Southwest Airlines (~398 USD).